examples/07_end_to_end/earthquake.ipynb

```diff
@@ -0,0 +1,416 @@
+    "    data,\n",
+    "    plugin=Plugin.MAP_SCATTER,\n",
+    "    columns=[\"longitude\", \"latitude\", \"magnitude\", \"magnitude\", \"time\", \"creation_time\"],\n",
+    "    #tooltip=[\"magnitude\", \"time\"]\n",
+    ")\n",
+    "# Create a tab widget with some PerspectiveWidgets inside\n",
+    "widget = widgets.Tab()\n",
+    "widget.children = [datagrid, worldmap]\n",
+    "widget.titles = [\"All events\", \"World map\"]\n",
+    "widget"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "e249cf63-355e-4e65-bd9c-d475cbc2d310",
+   "metadata": {},
+   "source": []
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "0eb12869-eb31-4eba-8079-5fa6ba28f892",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# PushPullAdapter\n",
+    "import threading\n",
+    "import csp\n",
+    "from obspy import read_events\n",
+    "from datetime import datetime, timedelta\n",
+    "from csp.impl.pushpulladapter import PushPullInputAdapter\n",
+    "from csp.impl.wiring import py_pushpull_adapter_def\n",
+    "from obspy.core.utcdatetime import UTCDateTime\n",
+    "import time\n",
+    "\n",
+    "class EventData(csp.Struct):\n",
+    "    time: UTCDateTime\n",
+    "    longitude: float\n",
+    "    latitude: float\n",
+    "    magnitude: float\n",
+    "    creation_time: datetime\n",
+    "\n",
+    "# Create a runtime implementation of the adapter\n",
+    "class FetchEventDataAdapter(PushPullInputAdapter):\n",
+    "    def __init__(self, interval, url, realtime):\n",
+    "        self._interval = interval\n",
+    "        self._thread = None\n",
+    "        self._running = False\n",
+    "        self._url = url\n",
+    "        self._realtime = realtime\n",
+    "\n",
+    "    def start(self, starttime, endtime):\n",
+    "        print(\"FetchEventDataAdapter::start\")\n",
+    "        self._running = True\n",
+    "        self._thread = threading.Thread(target=self._run)\n",
+    "        self._thread.start()\n",
+    "        self._starttime = datetime.utcnow()\n",
+    "\n",
+    "    def stop(self):\n",
+    "        print(\"FetchEventDataAdapter::stop\")\n",
+    "        if self._running:\n",
+    "            self._running = False\n",
+    "            self._thread.join()\n",
+    "\n",
+    "    def _run(self):\n",
+    "        history = []\n",
+    "        while self._running:\n",
+    "            now = datetime.utcnow()\n",
+    "            print(\"----------------------------------------------\")\n",
+    "            print(f\"{now}: refreshing earthquake feed\")\n",
+    "            print(\"----------------------------------------------\")\n",
+    "            catalog = read_events(self._url, format=\"QUAKEML\")\n",
+    "            def sorting(event):\n",
+    "                return event.origins[0].time\n",
+    "            catalog.events.sort(key=sorting)\n",
+    "            # tick whenever new events appear\n",
+    "            live = self._realtime\n",
```

Comment:
so this example is really missing the point of PushPull.  You shouldnt pass realtime as a flag at all into this adapter.
this code should grab all data from starttime ( drop data before starttime ) and call push_tick on it
as soon as it gets an event that is > utcnow() it should start passing live = True.
Note that you can only go from !live -> live and only once, so wneed to make sure liveliness sticks.  We should never push out of order events in any case.
While catching up there is no need to sleep at all, only sleep once we are "caught up" and pushing realtime